nmstatectl/README.varlink.md

```diff
@@ -0,0 +1,158 @@
+
+
+## Initiate nmstate varlink service
+User can initate nmstate-varlink services using ```nmstatectl``` command-line tool by defining the unix socket address path. The current nmstate-varlink implementation limits only using unix sockets connection.
+
+unix socket connection
+```bash
+$ nmstatectl varlink /run/nmstate.so &
+```
+
+Nmstate-varlink systemd services default listening address is ```unix:/run/nmstate/nmstate.so```
+systemd service initiation
+```bash
+$ systemctl start nmstate-varlink.service
+```
+
+After the `io.nmstate` interface is resolved using varlink resolver. Methods can be called directly via stdin/out without specifying the connection.
+`varlink call io.nmstate.Show`
+
+## Nmstate basic operations using varlink
+The basic functions from libnmstate (show, apply, commit and rollback) are called via varlink interface. Passing inputs to the functions only support JSON format. Below are the examples for each basic functions using varlink stdin/out and varlink client.
+
+Using libnmstate show function via varlink (query network state)
+
+Required for varlink stdin/out operation
+```bash
+$ dnf install libvarlink-utils
+```
+
+Varlink stdin/out:
+```bash
+$ varlink call unix:/run/nmstate.so/io.nmstate.Show
+
+# If interface is resolved using the varlink resolver
+
+$ varlink call io.nmstate.Show
+```
+
+Varlink python client:
+```python
+import varlink
+
+with varlink.Client("unix:/run/nmstate.so").open("io.nmstate") as nmstate:
+    nmstate.Show()
+```
+
+JSON output: libnmstate current network is reported under "state" object.
+```json
+{
+  "log": [
+    {
+      "level": "DEBUG",
+      "message": "Async action: Retrieve applied config: foo started",
+      "time": "2020-08-05 10:22:19"
+    },
+    {
+      "level": "DEBUG",
+      "message": "Async action: Retrieve applied config: foo finished",
+      "time": "2020-08-05 10:22:19"
+    }
+
+  ],
+  "state": {
+    "dns-resolver": {
+      "config": {
+        "search": [],
+        "server": []
+      },
+      "running": {}
+    },
+    "interfaces": [
+      {
+        "ipv4": {
+          "enabled": false
+        },
+        "ipv6": {
+          "enabled": false
+        },
+        "lldp": {
+          "enabled": false
+        },
+        "mac-address": "36:66:98:1D:6A:C8",
+        "mtu": 1500,
+        "name": "eth0",
+        "state": "down",
+        "type": "ethernet"
+      },
+    ],
+    "route-rules": {
+      "config": []
+    },
+    "routes": {
+      "config": [],
+      "running": []
+    }
+  }
+}
+```
+
+Using libnmstate apply function via varlink (query network state)
+
+Varlink stdin/out:
+state should be passed in json string format and varlink doesn't support passing json file.
+```bash
+$ varlink call unix:/run/nmstate.so/io.nmstate.Apply '{"arguments": {"desired_state": {"interfaces": [{"name": "foo", "type": "dummy", "state": "up", "ipv4": {"enabled": false}, "ipv6": {"enabled": false}}]} } }'
```

Comment:
We don't need to explain how varlink itself works which it might change in the future.

Please keep the document only contains the varlink interface nmstate provides.